ledger/src/ui/src/communication.c

```diff
@@ -22,44 +22,89 @@
  * IN THE SOFTWARE.
  */
 
-#include "apdu.h"
-#include "defs.h"
 #include "os.h"
+#include "apdu.h"
 #include "communication.h"
+#include "modes.h"
 
-/*
+/**
  * Implement the RSK ECHO command.
  *
  * @arg[in]  rx          number of received bytes from the Host
- * @ret                  number of transmited bytes to the host
+ * @returns              number of transmited bytes to the host
  */
 unsigned int echo(unsigned int rx) {
     return rx;
 }
 
-/*
+/**
  * Implement the RSK MODE command.
  *
- * Since this is only ever called from the bootloader, this always returns
- * RSK_MODE_BOOTLOADER
+ * This returns either bootloader or heartbeat mode, depending
+ * on the argument (which signals where it is called from)
  *
- * @ret number of transmited bytes to the host
+ * @arg[in] ui_heartbeat_main whether called from the ui heartbeat main
+ * @returns number of transmited bytes to the host
  */
-unsigned int get_mode() {
+unsigned int get_mode(bool ui_heartbeat_main) {
     unsigned char output_index = CMDPOS;
```

Comment:
This function always returns either `APP_MODE_UI_HEARTBEAT` or `APP_MODE_BOOTLOADER`, depending on where this was called from. This logic of a function having different behaviors depending on where it was called from is confusing. Also, it is hard to understand what `get_mode(false)` actually means just by looking at the code. What do you think of breaking this into two different functions, something along the lines of:

```
unsigned int report_mode_bootloader()
unsigned int report_mode_heartbeat()
```

Reply:
Good point, fixed!